server/engine/src/main/java/org/eclipse/lsp/cobol/implicitDialects/cics/utility/CICSOptionsCheckBaseUtility.java

```diff
@@ -32,391 +33,537 @@
+        throwException(
+                ErrorSeverity.ERROR, getLocality(rule), "Invalid option or parameter provided: ", message);
     }
-  }
-
-  protected void getAllTokenChildren(
-      ParserRuleContext ctx, List<TerminalNode> children, boolean validateResponseHandler) {
-    if (ctx.children == null) return;
-    ctx.children.forEach(
-        child -> {
-          if (TerminalNode.class.isAssignableFrom(child.getClass())
-                  && baseDuplicateOptions.containsKey(((TerminalNode) child).getSymbol().getType()))
-            children.add((TerminalNode) child);
-          else if (ParserRuleContext.class.isAssignableFrom(child.getClass())) {
-            if (validateResponseHandler
-                    && child.getClass().getSimpleName().equals("Cics_handle_responseContext"))
-              checkResponseHandlers((CICSParser.Cics_handle_responseContext) child);
-            if (!CICSParser.Cics_data_areaContext.class.isAssignableFrom(child.getClass())
-                    && !CICSParser.Cics_nameContext.class.isAssignableFrom(child.getClass())
-                    && !CICSParser.Cics_data_valueContext.class.isAssignableFrom(child.getClass())) {
-              getAllTokenChildren((ParserRuleContext) child, children, validateResponseHandler);
+
+    /**
+     * Validates parser rules to ensure browser functionality
+     *
+     * @param ctx            Context to evaluate
+     * @param coreTokenIndex The CICSParser index(ices) for the core token(s)
+     */
+    public void checkBrowsingInvalidOptions(ParserRuleContext ctx, int... coreTokenIndex) {
+        if (ctx.children == null) return;
+        for (int index = 0; index < ctx.children.size(); index++) {
+            if (!TerminalNode.class.isAssignableFrom(ctx.children.get(index).getClass())) continue;
+            int tokenIndex = ((TerminalNode) ctx.children.get(index)).getSymbol().getType();
+            boolean isCoreToken = ArrayUtils.contains(coreTokenIndex, tokenIndex);
+            if (!isCoreToken
+                    && tokenIndex != CICSLexer.LPARENCHAR
+                    && tokenIndex != CICSParser.START
+                    && tokenIndex != CICSParser.AT
+                    && tokenIndex != CICSParser.END
+                    && tokenIndex != CICSParser.NEXT
+                    && tokenIndex != CICSParser.RULE_cics_handle_response) {
+                throwBrowsingViolation(
+                        ctx.children.get(index),
+                        "Accessory options not allowed when browsing with START or END");
+            } else if (isCoreToken
+                    && index + 1 < ctx.children.size()
+                    && ctx.children.get(index + 1).getText().contains("(")) {
+                if (TerminalNode.class.isAssignableFrom(ctx.children.get(index + 1).getChild(0).getClass()) && ((TerminalNode) ctx.children.get(index + 1).getChild(0)).getSymbol().getType() == CICSLexer.LPARENCHAR) {
+                    index++;
+                }
```

Comment:
I am not sure I follow the logic here. We try to skip a rule (I assume) consisting of `(...)` and if it happens to be just the opening paren, we (I assume) skip the content, which means we are now at the right paren and we will fail in the next round. Or am I misunderstanding this?